### ISSUE TITLE
Segmentation fault when calling FlannBasedMatcher.train()

### ISSUE THREAD

**Issue by JustASquid** [opencv/opencv#17448]
##### System information (version)

- OpenCV => Tried 3.4.2, 4.2.0
- Operating System / Platform => Windows 64 Bit, Ubuntu 64 Bit (Tried both)
- Compiler => N/A (using prebuilt opencv from Conda distribution, also tried pip opencv-python)

##### Detailed description

Using a FlannBasedMatcher, the process crashes with exit code `0xC0000005` when `matcher.train()` is called after adding keypoint descriptors with `matcher.add()`

Note - I know the listed versions of OpenCV I tried with are not the latest opencv versions, but they are the latest prebuilt binaries offered by Python package managers. Building from source turned out to be quite difficult so I wasn't able to fully follow that thread, but if anyone can confirm that this was fixed in a recent version I'm happy to close this and ask the Python package maintainers to update their binaries.

##### Steps to reproduce

```python
import cv2

# Content of image doesn't matter, so long as it contains enough info for some keypoints to be found
image = cv2.imread('test.jpg')

detector = cv2.ORB_create()
keypoints, descriptors = detector.detectAndCompute(image, mask=None)

# Parameters from
# https://opencv-python-tutroals.readthedocs.io/en/latest/py_tutorials/py_feature2d/py_matcher/py_matcher.html
FLANN_INDEX_LSH = 6
index_params= dict(algorithm = FLANN_INDEX_LSH,
                   table_number = 6, # 12
                   key_size = 12,     # 20
                   multi_probe_level = 1) #2
search_params = dict(checks=50)   # or pass empty dictionary

matcher = cv2.FlannBasedMatcher(index_params,search_params)


matcher.add(descriptors)

# Crash occurs here
matcher.train()


print("Finished")
```

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   
 - [x] I checked the problem with documentation, FAQ, open issues,
       answers.opencv.org, Stack Overflow, etc and have not found solution
   
 - [x] I updated to latest OpenCV version and the issue is still there
   
 - [x] There is reproducer code and related data files: videos, images, onnx, etc

**Comment by asmorkalov** [comment]
The issue reproduced with Ubuntu 18.04 and current 3.4 branch. The fault happens in `flann::Index` constructor in line https://github.com/opencv/opencv/blob/3.4/modules/features2d/src/matchers.cpp#L1166

**Comment by JustASquid** [comment]
@asmorkalov Thanks for confirming.
Can you suggest any workarounds?

**Comment by asmorkalov** [comment]
@JustASquid Not now. I suspect some bug in python bindings, but it should be checked.

**Comment by alalek** [comment]
Perhaps relates to this issue: #17553
Please check this fix for master (4.5.1+): #19045
